test/e2e/storage/testsuites/provisioning.go

```diff
@@ -76,183 +79,156 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
-			if needsCleanup {
-				resource.cleanupResource(driver, pattern)
-			}
-		})
-
-		// Ginkgo's "Global Shared Behaviors" require arguments for a shared function
-		// to be a single struct and to be passed as a pointer.
-		// Please see https://onsi.github.io/ginkgo/#global-shared-behaviors for details.
-		testProvisioning(&input)
+func (p *provisioningTestSuite) defineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	var (
+		dInfo       = driver.GetDriverInfo()
+		dDriver     DynamicPVTestDriver
+		config      *PerTestConfig
+		testCleanup func()
+		testCase    *StorageClassTest
+		cs          clientset.Interface
+		pvc         *v1.PersistentVolumeClaim
+		sc          *storage.StorageClass
+	)
+
+	BeforeEach(func() {
+		// Check preconditions.
+		if pattern.VolType != testpatterns.DynamicPV {
+			framework.Skipf("Suite %q does not support %v", p.tsInfo.name, pattern.VolType)
+		}
+		ok := false
+		dDriver, ok = driver.(DynamicPVTestDriver)
+		if !ok {
+			framework.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+		}
 	})
-}
 
-type provisioningTestResource struct {
-	driver TestDriver
-
-	claimSize string
-	sc        *storage.StorageClass
-	pvc       *v1.PersistentVolumeClaim
-	// follow parameter is used to test provision volume from snapshot
-	vsc *unstructured.Unstructured
-}
-
-var _ TestResource = &provisioningTestResource{}
-
-func (p *provisioningTestResource) setupResource(driver TestDriver, pattern testpatterns.TestPattern) {
-	// Setup provisioningTest resource
-	switch pattern.VolType {
-	case testpatterns.DynamicPV:
-		if dDriver, ok := driver.(DynamicPVTestDriver); ok {
-			p.sc = dDriver.GetDynamicProvisionStorageClass("")
-			if p.sc == nil {
-				framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", driver.GetDriverInfo().Name)
-			}
-			p.driver = driver
-			p.claimSize = dDriver.GetClaimSize()
-			p.pvc = getClaim(p.claimSize, driver.GetDriverInfo().Config.Framework.Namespace.Name)
-			p.pvc.Spec.StorageClassName = &p.sc.Name
-			framework.Logf("In creating storage class object and pvc object for driver - sc: %v, pvc: %v", p.sc, p.pvc)
-			if sDriver, ok := driver.(SnapshottableTestDriver); ok {
-				p.vsc = sDriver.GetSnapshotClass()
-			}
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("provisioning")
+
+	init := func() {
+		// Now do the more expensive test initialization.
+		config, testCleanup = driver.PrepareTest(f)
+		cs = config.Framework.ClientSet
+		claimSize := dDriver.GetClaimSize()
+		sc = dDriver.GetDynamicProvisionStorageClass(config, "")
+		if sc == nil {
+			framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
```

Comment:
these Skip messages should specify that the sc is not defined because the fstype %s is not supported. there is no other reason (currently?) for a driver to implement DynamicPVTestDriver but return nil

Reply:
Can this code here know why the driver didn't return a storage class? I don't think it should make assumptions based on the current implementation.

Reply:
yeah I guess it's fine as-is. I guess it is the driver author's responsibility to make sense of why their GetDynamicProvisionStorageClass would return nil. Maybe it could return an error in the future, seems pointless now though